File: Ruby/BasicAlgorithms/FindMaximumValue.rb
# Write a function that takes an integer array and prints and returns the maximum value in the array. 
# Your program should also work with a given array that has all negative numbers 
# input:  [-3, -5, -7] 
# output: -3

module FindMaximumValue

  def FindMaximumValue.For(numArray)
    max = 0
    for num in numArray
      if max < num 
        max = num
      end
    end
    print "For Max Number: ", max, "\n"
    return max
  end
  
  def FindMaximumValue.While(numArray)
    iterator = 0
    max = 0
    begin
      if max < numArray[iterator]
        max = numArray[iterator]
      end
      iterator += 1    
    end while iterator < numArray.length
    print "While Max Number: ", max, "\n"
    return max
  end

  def FindMaximumValue.DoUntil(numArray)
    iterator = 0
    max = 0
    until iterator == numArray.length do
      if max < numArray[iterator]
        max = numArray[iterator]
      end
      iterator += 1      
    end
    print "Until Max Number: ", max, "\n"
    return max
  end
  
end

File: CSharp/BasicAlgorithms/FindMaximumValue.cs
using System;

namespace CSharpBasics;

/* 
* Write a function that takes an integer array and prints and returns the maximum value in the array. 
* Your program should also work with a given array that has all negative numbers (e.g. [-3, -5, -7]), 
* or even a mix of positive numbers, negative numbers and zero.
*/

public static partial class BasicAlgorithms
{
    public static int FindMaximumValue_For(int[] integerArray)
    {
        int maximumValue = integerArray[0];
        for (int iterator = 0; iterator < integerArray.Length; iterator++)
        {
            if(maximumValue < integerArray[iterator])
            {
                maximumValue = integerArray[iterator];
            }    
        }
        Console.WriteLine($"~~~ Find Maximum Value In Array -- For Loop -- Value: {maximumValue} ~~~");
        return maximumValue;
    }
    public static int FindMaximumValue_Do_While(int[] integerArray)
    {
        int iterator = 0;
        int maximumValue = integerArray[0];
        do 
        {
            if(maximumValue < integerArray[iterator])
            {
                maximumValue = integerArray[iterator];
            }   
            iterator++;
        } 
        while (iterator < integerArray.Length);

        Console.WriteLine($"~~~ Find Maximum Value In Array -- Do While Loop -- Value: {maximumValue} ~~~");
        return maximumValue;
    }
    public static int FindMaximumValue_While(int[] integerArray)
    {
        int iterator = 0;
        int maximumValue = integerArray[0];
        while (iterator < integerArray.Length)
        {
            if(maximumValue < integerArray[iterator])
            {
                maximumValue = integerArray[iterator];
            }  
            iterator++;
        }

        Console.WriteLine($"~~~ Find Maximum Value In Array -- While Loop -- Value: {maximumValue} ~~~");
        return maximumValue;
    }
}